Samples/ios/SampleApp/SampleAppiOS/SampleMSALUtil.m

```diff
@@ -0,0 +1,217 @@
+    NSString *currentUserIdentifier = [self currentUserIdentifer];
+    if (!currentUserIdentifier)
+    {
+        // If we did not find an identifier then return nil with a no user signed
+        // in error specific for the application.
+        if (error)
+        {
+            *error = [NSError errorWithDomain:SampleAppErrorDomain code:SampleAppNoUserSignedIn userInfo:nil];
+        }
+        
+        return nil;
+    }
+    
+    
+    // Because error is an optional parameter we need to pass in our own error pointer to make sure we get
+    // an error back so we can inspect it after.
+    NSError *localError = nil;
+    
+    // Ask MSALPublicClientApplication object to retrieve the user from the cache.
+    MSALUser *user = [[self clientApplication] userForIdentifier:currentUserIdentifier error:&localError];
+    
+    // If we did not find a user because it wasn't found in the cache then that must mean someone else removed
+    // the user underneath us, either due to multiple apps sharing a client ID, or due to the user restoring an
+    // image from another device. In this case it is best to detect that case and clean up local state.
+    if (!user && [localError.domain isEqualToString:MSALErrorDomain] && localError.code == MSALErrorUserNotFound)
+    {
+        [self cleanupLocalState];
+    }
+    
+    // NSError ** is traditionally an optional parameters, so we should check to make sure it is not nil before
+    // filling it.
+    if (error)
+    {
+        *error = localError;
+    }
+    
+    return user;
+}
+
++ (void)signInUser:(void (^)(MSALUser *user, NSString *token, NSError *error))signInBlock
+{
+    MSALPublicClientApplication *application = [self clientApplication];
+    
+    // When signing in a user for the first time we acquire a token without providing
+    // a user object. If you've previously asked the user for an email address,
+    // or phone number you can provide that as a "login hint."
+    [application acquireTokenForScopes:@[@"User.Read"]
+                       completionBlock:^(MSALResult *result, NSError *error)
+    {
+        if (error)
+        {
+            signInBlock(nil, nil, error);
+            return;
+        }
+        
+        // In the initial acquire token call we'll want to look at the user object
+        // that comes back in the result.
+        MSALUser *user = result.user;
+        
+        // The userIdentifier in the MSALUser is the key to retrieve this user from
+        // the cache in the future. Save this piece of information in a place you can
+        // easily retrieve in your app. In this case we're going to store it in
+        // NSUserDefaults.
+        [[NSUserDefaults standardUserDefaults] setValue:user.userIdentifier forKey:CURRENT_USER_KEY];
+        
+        signInBlock(user, result.accessToken, error);
+    }];
+}
+
++ (void)acquireTokenSilentForCurrentUser:(void (^)(NSString *token, NSError *error))acquireTokenBlock
```

Comment:
Apple rarely names completion block types in their own headers, usually leaving them declared like this.